File: ffi/minidraw/minidraw.cpp
#include "minidraw.h"
#include <cstdio>

drawBoard* a;

void initDrawBoard(int w, int h)
{
    if (a == nullptr) {
        a = new drawBoard();
    }
    a->init(w, h);
};

void savePng(char* name)
{
    a->savePng(name);
}

void setColor(int x, int y, int z)
{
    a->setColor(x, y, z);
}

void lineTo(int x, int y)
{
    a->lineTo(x, y);
}

void moveTo(int x, int y)
{
    a->moveTo(x, y);
}

void stroke()
{
    a->stroke();
};

bool drawBoard::init(int64_t width, int64_t height)
{
    cairo_destroy(cr);
    cairo_surface_destroy(surface);

    surface = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, width, height);
    cr = cairo_create(surface);
    cairo_set_source_rgb(cr, 1, 1, 1);
    cairo_paint(cr);
    return true;
}

void drawBoard::moveTo(double x, double y)
{
    cairo_move_to(cr, x, y);
}

void drawBoard::lineTo(double x, double y)
{
    cairo_line_to(cr, x, y);
}

void drawBoard::stroke()
{
    cairo_stroke(cr);
}

void drawBoard::savePng(char* name)
{
    auto status = cairo_surface_write_to_png(surface, name);
    printf("%d -> %s\n", status, name);
}

void drawBoard::setColor(int x, int y, int z)
{
    cairo_set_source_rgb(cr, x / 255, y / 255, z / 255);
}

File: ffi/julia/minidraw.jl
module minidraw

libp = "../minidraw/libminidraw.so"

function version()
    unsafe_string(@ccall libp.version()::Cstring)
end

function initDrawBoard(x,y)
    @ccall libp.initDrawBoard(x::Cint,y::Cint)::Cvoid
end

function setColor(r,g,b)
    @ccall libp.setColor(r::Cint,g::Cint,b::Cint)::Cvoid
end

function lineTo(x,y)
    @ccall libp.lineTo(x::Cint,y::Cint)::Cvoid
end

function moveTo(x,y)
    @ccall libp.moveTo(x::Cint,y::Cint)::Cvoid
end

function stroke()
    @ccall libp.stroke()::Cvoid
end

function savePng(name)
    @ccall libp.savePng(name::Cstring)::Cvoid
end

end